### PR TITLE
Delete cookie from browser if session becomes empty

### PR DESCRIPTION
**Pull request by PowerKiKi**
|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | yes
| BC Break      | no
| New Feature   | yes
| RFC           | no
| QA            | no

### Description

In a logout process this would allow us to user `$session->clear();` in
order to delete the session on the server side and now also delete the
cookie on the browser side.

While in most cases it's not critical to have leftover cookies on the browser
side, it is not a good practice. And it can become a real problem if use-case
where the cookie existence is used as a condition for some behavior, for
instance redirections via nginx configuration.

I am not quite sure if this should be considered a bugfix or a new feature.

### CODE REVIEW

File: src/PhpSessionPersistence.php [hunk 1]
```diff
@@ -370,6 +370,10 @@ private function responseAlreadyHasCacheHeaders(ResponseInterface $response) : b
 
     private function getCookieLifetime(SessionInterface $session) : int
     {
+        if (empty($session->toArray())) {
```
Comment (samsonasik) [inline src/PhpSessionPersistence.php]: can use truthy check: 

```suggestion
        if (! $session->toArray()) {
```

Comment (PowerKiKi) [reply src/PhpSessionPersistence.php]: I was under the impression that truthy check for an array was against laminas code style. Am I confusing with another project ?

Comment (samsonasik) [reply src/PhpSessionPersistence.php]: previously, I proposed pr on some repos for eliminate count() for empty check on arrays, with empty() check, and @Ocramius review it to use truthy check instead. The changes already applied on some repos for that. I think that can be used for consistency.

Comment (PowerKiKi) [reply src/PhpSessionPersistence.php]: Thank for letting me know. This is also my preferred style, so I'm glad it's accepted.

**Comment by samsonasik** [comment]
I talked with @boesing that it may need to be configurable. Maybe this should be configured the same way as non_locking flag?

The new config may be : `$deleteCookieOnEmptySession = false` or `$clearCookieOnEmptySession = false`.

**Comment by PowerKiKi** [comment]
What would be the real world use-case to keep the cookie for a session that is empty ?

AFAIK there is nothing in `SessionInterface` that could inform us whether the session has been loaded from an existing cookie, or if the session has just been created in memory and was never persisted at all.

So an empty session cannot be used to somehow decide something about the client. Whether it is an "already known" client or a new one. The only way to do that is to actually put some kind of value into the session, maybe a user ID, or juste a `beenHereBefore` boolean flag.

TL;DR: an empty session is useless and thus has no reason to be persisted

**Comment by Ocramius** [comment]
PHP is not the only consumer of cookies BTW: deleting cookies can lead to a lot of "fun" side-effects, especially when talking about weird WAF that corporates use.

> So an empty session cannot be used to somehow decide something about the client. Whether it is an "already known" client or a new one. 

the session: no.

the cookie: yes, it is an identifier after all.

I don't mind either way, but careful with dropping information implicitly 😅

**Comment by PowerKiKi** [comment]
> PHP is not the only consumer of cookies

It is precisely the reason I came up with this PR. My nginx will serve different things based on the presence of the session cookie. When a user logs out, I need the cookie to be entirely removed so that nginx can serve the "anonymous" version of the website.

I imagine this would also help for caching proxy. As long as the user is logged you probably shouldn't serve him cached pages, but as soon as he logs out and is "anonymous" again you should be able to serve him cached pages.

I don't really see this as dropping information implicitly. You explicitly have to call `$session->clear()` to empty session and now also delete cookie. If you need to log out the user but for some reason keep his session, you can do something along the lines of:

```php
$session->set('user', null);
$session->set('visitedThisWebsiteOnceBefore', true);
```

It all seems coherent to me.

Did you encounter real world cases where the existence of a cookie for a PHP session (not just any cookie), whose session was empty, was indeed a key element in WAF ? What kind of feature would be broken if that "empty session cookie" was removed ?

**Comment by PowerKiKi** [comment]
@samsonasik, I didn't see any concrete case where this PR would create problems in the last month. So could we reconsider to merge this as is ?

**Comment by samsonasik** [comment]
@PowerKiKi like previously stated, it requires a new config so people that want to keep the cookie will still can work with this. 

The use case may be an after logged out, a user still has a "customer ID" cookie to show the suggested product.

**Comment by PowerKiKi** [comment]
I added configuration, docs and changelog. Please review again when you have time.

**Review by samsonasik — APPROVED** [review]

File: docs/book/intro.md [hunk 2]
```diff
@@ -53,6 +53,33 @@ that the session data won't be changed or when only one of the concurrent script
...
+            ],
+        ],
+    ],
+];
+```
+
+Then whenever you call `$session->clear();`, the cookie will be deleted from in
+the browser in addition to deleting session on the server.
+
+This is the recommended configuration to prevent leftover in the browser, unless
```
Comment (samsonasik) [inline docs/book/intro.md]: I am not sure make recommendation to remove cookie on empty session is ok. It is always depends on what developer want to achieve.

Comment (PowerKiKi) [reply docs/book/intro.md]: I removed the recommendation.

File: docs/book/intro.md [hunk 3]
```diff
@@ -53,6 +53,30 @@ that the session data won't be changed or when only one of the concurrent script
...
+### Enabling deletion of cookie in browser
+
+The default behaviour of the php session extension is to never delete the cookie
+in the browser. It is possible to automatically delete the cookie in the browser
+when the session becomes empty.
+
+This option can be enabled using the following configuration:
+
+```php
+// file: data/session.global.php
```
Comment (samsonasik) [inline docs/book/intro.md]: I think it should be : 

```suggestion
// file: config/autoload/session.global.php
```

Comment (PowerKiKi) [reply docs/book/intro.md]: I did for both examples.

File: src/PhpSessionPersistence.php [hunk 4]
```diff
@@ -125,6 +134,15 @@ public function isNonLocking() : bool
         return $this->nonLocking;
     }
 
+    /**
+     * @internal
```
Comment (weierophinney) [inline src/PhpSessionPersistence.php:138]: Private methods do not need to be marked `@internal`; they're implicitly internal, as they cannot be called by any but the defining class.

(I'm currently rebasing your patch against develop, and will make this change before pushing.)

**Review by weierophinney — APPROVED** [review]
I've changed the base to develop, as this is a new feature and needs to target a new minor version; to facilitate getting it in, I've also rebased and pushed, with some additional commits around documentation.

:ship:
